File: Union_Find/547.Friend-Circles/547.Friend-Circles.java
class Solution {
    int[] parents;
    int[] size;
    int count;
    public int findCircleNum(int[][] isConnected) {
        int n = isConnected.length;
        parents = new int[n];
        size = new int[n];
        count = n;
        for (int i = 0; i < n; i++) {
            parents[i] = i;
            size[i] = 1;
        }

        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                if (i == j || isConnected[i][j] == 0) {
                    continue;
                }
                union(i, j);
            }
        }
        return count;
    }
    
    private int find(int p) {
        while (p != parents[p]) {
            p = parents[p];
        }
        return p;
    }

    private void union(int p, int q) {
        int pRoot = find(p);
        int qRoot = find(q);
        
        if (pRoot == qRoot) {
            return;
        }

        if (size[pRoot] < size[qRoot]) {
            parents[pRoot] = qRoot;
            size[qRoot] += size[pRoot];
        } else {
            parents[qRoot] = pRoot;
            size[pRoot] += size[qRoot];
        }
        count -= 1;
    }
}

File: Union_Find/547.Friend-Circles/547.Friend-Circles.py
class Solution(object):
    
    def findCircleNum(self, isConnected):
        """
        :type isConnected: List[List[int]]
        :rtype: int
        """
        n = len(isConnected)
        parents = [i for i in range(0, n)]
        size = [1] * n
        count = n

        def find(p):
            while p != parents[p]:
                p = parents[p]
            return p

        def union(p, q):
            nonlocal count
            pRoot = find(p)
            qRoot = find(q)

            if (pRoot == qRoot):
                return

            if size[pRoot] < size[qRoot]:
                parents[pRoot] = qRoot
                size[qRoot] += size[pRoot]
            else:
                parents[qRoot] = pRoot
                size[pRoot] += size[qRoot]
            count -= 1

        for i in range(0, n):
            for j in range(i, n):
                if i == j or isConnected[i][j] == 0:
                    continue
                union(i, j)
        return count
